### PR TITLE
Upgrade to Qt 5.15.0

### LINKED ISSUE #180: Upgrade to Qt 5.15 LTS
Enhancement Request
--------------
Qt 5.11 is used currently, which isn't a LTS
Qt 5.12 is a LTS, and Qt 5.15 LTS was released in May, might be a good idea to upgrade to Qt 5.15 LTS directly.

https://www.qt.io/blog/qt-5.15-released

### PR DESCRIPTION
**Pull request by jmarrec**
If it works Fixes #180 and perhaps it will also fix the issue #203

### CODE REVIEW

**Comment by jmarrec** [comment]
Some warnings at config stage.

```
 CMake Warning (dev) at /home/julien/Software/Others/OSApp-build-release/Qt-install/5.15.0/gcc_64/lib/cmake/Qt5Core/Qt5CoreMacros.cmake:44 (message):
   qt5_get_moc_flags is not part of the official API, and might be removed in
   Qt 6.
 Call Stack (most recent call first):
   /home/julien/Software/Others/OSApp-build-release/Qt-install/5.15.0/gcc_64/lib/cmake/Qt5Core/Qt5CoreMacros.cmake:86 (_qt5_warn_deprecated)
   ProjectMacros.cmake:92 (qt5_get_moc_flags)
   src/model_editor/CMakeLists.txt:116 (qt5_wrap_cpp_minimally)  
 This warning is for project developers.  Use -Wno-dev to suppress it.
```

Otherwise it seems to locate the libs: 
image

**Comment by jmarrec** [comment]
I can build locally on Ubuntu. The CI is tripping on a dependency thing:

```
 E: Failed to fetch http://azure.archive.ubuntu.com/ubuntu/pool/main/libx/libx11/libx11-xcb-dev_1.6.4-3ubuntu0.2_amd64.deb  404  Not Found [IP: 52.154.174.208 80]
Fetched 1618 kB in 0s (11.2 MB/s)
E: Unable to fetch some archives, maybe run apt-get update or try with --fix-missing?
```

We have a bunch of deprecated functions, that was kinda expected.

**Comment by jmarrec** [comment]
@macumber qtwinmigrate has to be updated (or treat warning as errors removed for that one)

```
D:\a\OpenStudioApplication\OpenStudioApplication\src\qtwinmigrate\qwinhost.cpp(88,20): error C2220: the following warning is treated as an error [D:\a\OpenStudioApplication\OpenStudioApplication\build\src\qtwinmigrate\qtwinmigrate.vcxproj]
D:\a\OpenStudioApplication\OpenStudioApplication\src\qtwinmigrate\qwinhost.cpp(88,20): warning C4996: 'Qt::WA_NoBackground': was declared deprecated [D:\a\OpenStudioApplication\OpenStudioApplication\build\src\qtwinmigrate\qtwinmigrate.vcxproj]
```

Is this from https://github.com/qtproject/qt-solutions/tree/master/qtwinmigrate? Is it really needed?

**Comment by macumber** [comment]
The 3D view model tab did not work for me after updating, I was able to get it working with:

```
MacBook-Pro-2:build mymac$ export QTWEBENGINE_CHROMIUM_FLAGS="--ignore-gpu-blacklist"
MacBook-Pro-2:build mymac$ /Applications/OpenStudioApplication-1.1.0-alpha/OpenStudioApp.app/Contents/MacOS/OpenStudioApp 

DevTools listening on ws://127.0.0.1:52139/devtools/browser/bbd7a4c8-574d-47f8-995a-3773f3db505f
Unable to create basic Accelerated OpenGL renderer.
Unable to create basic Accelerated OpenGL renderer.
Core Image is now using the software OpenGL renderer. This will be slow.
Unable to create basic Accelerated OpenGL renderer.
Unable to create basic Accelerated OpenGL renderer.
Core Image is now using the software OpenGL renderer. This will be slow.
Unable to create basic Accelerated OpenGL renderer.
Unable to create basic Accelerated OpenGL renderer.
Core Image is now using the software OpenGL renderer. This will be slow.
Unable to create basic Accelerated OpenGL renderer.
Unable to create basic Accelerated OpenGL renderer.
Core Image is now using the software OpenGL renderer. This will be slow.
```

I think we probably need to update the ViewModel measure, my guess is that is too old.  We should disable the blacklist as a workaround.

**Comment by macumber** [comment]
Updating Qt did seem to change the dark mode behavior.  The App updates instantly when changing dark to light mode, no restart needed.  The App is usable in dark mode but doesn't look great:

Light Mode:
image

Dark Mode:
image

File: .github/workflows/app_build.yml [hunk 1]
```diff
@@ -272,12 +278,33 @@ jobs:
...
         # Below needed if you re-enable 'Download the OpenStudio installer' step
         # openstudio_DIR=$(find `pwd`/OpenStudio-$OS_SDK_VERSION/ -name "openstudioConfig.cmake")
         # openstudio_DIR=$(dirname "${openstudio_DIR}")
         # echo "openstudio_DIR=${openstudio_DIR}"
         # echo "::set-env name=openstudio_DIR::$openstudio_DIR"
+        echo "Install aqtinstall, then install Qt 5.15.0 for ${{ matrix.QT_OS_NAME }} ${{ matrix.QT_ARCH }}"
+        pip3 install aqtinstall
+        aqt install --outputdir ./build/Qt-install/ 5.15.0 ${{ matrix.QT_OS_NAME }} desktop ${{ matrix.QT_ARCH }} -m qtwebengine qtwebglplugin
+        if [ "$RUNNER_OS" == "Windows" ]; then
+          echo "::add-path::$(pwd)/build/Qt-install/5.15.0/msvc2019_64"
```
Comment (macumber) [inline .github/workflows/app_build.yml:292]:
```suggestion
          echo "::add-path::$(pwd)/build/Qt-install/5.15.0/msvc2019_64/bin"
```

File: .github/workflows/app_build.yml [hunk 2]
```diff
@@ -272,12 +278,33 @@ jobs:
...
+          echo "::set-env name=QT_INSTALL_DIR::$(pwd)/build/Qt-install/5.15.0/msvc2019_64"
+          ls ./build/Qt-install/5.15.0/
+          ls ./build/Qt-install/5.15.0/msvc2019_64 || true
+          ls ./build/Qt-install/5.15.0/msvc2019_64/lib || true
+          ls ./build/Qt-install/5.15.0/msvc2019_64/lib/cmake || true
+        else
+          echo "::add-path::$(pwd)./build/Qt-install/5.15.0/${{ matrix.QT_ARCH }}"
```
Comment (macumber) [inline .github/workflows/app_build.yml:299]:
```suggestion
          echo "::add-path::$(pwd)./build/Qt-install/5.15.0/${{ matrix.QT_ARCH }}/bin"
```

**Comment by macumber** [comment]
@jmarrec there are a bunch of warnings treated as errors on Windows, I will start looking at those

**Review by macumber — APPROVED** [review]
App and plugin work on Windows

**Comment by macumber** [comment]
@jmarrec I'm ready to merge but wanted to confirm with you because this is a large change.  I updated some of the code to not use deprecated methods (all except the XML parsing in AccessPolicy), I did not thoroughly test the affected dropdowns, etc.

File: .github/workflows/app_build.yml (hunk 1 above)
Comment (jmarrec) [reply .github/workflows/app_build.yml]: Why is bin/ needed on windows?

Comment (jmarrec) [reply .github/workflows/app_build.yml]: I didnt test on windows directly, but I find that weird. Cmake during the find_package calls will look into lib/cmake

Comment (jmarrec) [reply .github/workflows/app_build.yml]: @macumber

**Comment by jmarrec** [comment]
Did you upgrade qtwinmigrate manually or pulled the latesr github version?

**Comment by macumber** [comment]
I pulled latest (July 2018) qtwinmigrate from https://github.com/qtproject/qt-solutions/tree/master/qtwinmigrate

File: .github/workflows/app_build.yml (hunk 1 above)
Comment (macumber) [reply .github/workflows/app_build.yml]: I downloaded Qt 5.15 on my computer using aqtinstall.  All of the tools like `moc` and `qmake` as well as the Qt dlls are under `bin`.  I assumed that is what we needed to add to the path?

**Comment by macumber** [comment]
@jmarrec it seems we have some CI build failures possibly due to the caching: https://github.com/openstudiocoalition/OpenStudioApplication/pull/227/checks?check_run_id=1078433277

Caching seems complicated, should we just consider dropping caching and building clean each time?

**Review by jmarrec — COMMENTED** [review]
> I pulled latest (July 2018) qtwinmigrate from https://github.com/qtproject/qt-solutions/tree/master/qtwinmigrate

Should we consider doing something cleaner a bit? Some or all of the following:
* Adding some kind of README or at least mentionning the commit SHA you pulled in the commit
* Exclude the folder from clang-format. I don't think it makes a ton of sense formatting external stuff. 
* Move the qtwinmigrate folder at root/third_party/qtwinmgrate

**Comment by jmarrec** [comment]
> @jmarrec it seems we have some CI build failures possibly due to the caching: https://github.com/openstudiocoalition/OpenStudioApplication/pull/227/checks?check_run_id=1078433277
> 
> Caching seems complicated, should we just consider dropping caching and building clean each time?

I made a stupid typo in there, I'll just fix it.

File: .github/workflows/app_build.yml (hunk 1 above)
Comment (jmarrec) [reply .github/workflows/app_build.yml]: Yeah ok that makes sense. Actually I adding the path to root not thinking about binaries.

I thought about changing this to remove `NO_DEFAULT_PATHS` so it woul just pick up any installed (system) Qt if you do not supply QT_INSTALL_DIR explicitly: https://github.com/openstudiocoalition/OpenStudioApplication/blob/97ca38d046531ca27a7f058fdf275c1bb5767fb8/CMakeLists.txt#L475

**Comment by jmarrec** [comment]
We should probably test the installers on clean machines, make sure they work fine.

**Comment by jmarrec** [comment]
Couple of changes in the appearance on Ubuntu 18.04

Here is OSApp 1.1.0 with Qt 5.11.2

image



And here is this PR with Qt 5.15.0

image

File: src/model_editor/Application.hpp [hunk 3]
```diff
@@ -68,9 +68,8 @@ class ApplicationSingleton
   QWidget* sketchUpWidget();
 
   /// Process pending Qt events
-  /// returns true if some work was done
-  bool processEvents();
-  bool processEvents(int maxTime);
+  void processEvents();
```
[the remark this answers is not in the document]
Comment (macumber) [reply src/model_editor/Application.hpp:71]: The underlying method we needed to return bool [`hasPendingEvents`](https://doc.qt.io/qt-5/qcoreapplication-obsolete.html#hasPendingEvents) was deprecated.

**Comment by jmarrec** [comment]
Confirmed in a clean 18.04 (in addition to 20.04) that this is needed: `sudo apt install libxcb-xinerama0`

**Comment by jmarrec** [comment]
# Ubuntu 20.04 (Focal Fossa)

I tested the installer on a clean Ubuntu VM. I thought I'd kill two birds with one stone and test with 20.04 (Focal Fossa) instead of the supported 18.04 (bionic): providing some slight adjustments It works!

Geometry tab doesn't work though. The 3D view errors, while the Editor / Create new floor plan works fine. (edited) 

I can make it segfault by very quickly swtiching tabs. (Randomly clicking over and over again between the last 4 or 5 tabs)

I cannot run a simulation. FT happens Ok, but E+ crashes
```
**FATAL:DLL problem cause program termination
terminate called after throwing an instance of 'EnergyPlus::FatalError'
  what():  DLL problem cause program termination
Aborted (core dumped)
```


----

# Test in Ubuntu 18.04 (supported platform)

Test in a clean VM 18.04:

`sudo apt install libxcb-xinerama0`  is needed on 18.04 too

The 3D view (unlike in 20.04) works

Same `FATAL:DLL problem cause program termination` when running E+ sim

I can also make it crash by clicking randomly very quickly on the last 5 tabs (it takes like 30 clicks or so...)

**Comment by powskier** [comment]
Hello, I used the windows installer, set it up and ran (tried) a couple of models.  Out of the box I got an error about not finding the openstudio-standards.
 Here's a screen shot.
StandardsError091020

I can work through the standards gem error but wanted to report my initial condition.

**Comment by macumber** [comment]
@jmarrec do you know what the "##[error]Validation Failed" messages mean on the failures with the release tagging?  Do you want to separate the release tagging and just merge the Qt stuff?

**Comment by jmarrec** [comment]
@macumber Yeah I'll do that on another branch. You can ignore that.

Did you get a chance to test on Mac? That's the last thing needed to merge this PR

**Comment by macumber** [comment]
App on Mac looks good, was able to use floorspace, only error was on run tab:

```
Initializing workflow.
Processing OpenStudio Measures.
Translating the OpenStudio Model to EnergyPlus.
Processing EnergyPlus Measures.
Starting Simulation.
**FATAL:DLL problem cause program termination
libc++abi.dylib: terminating with uncaught exception of type EnergyPlus::FatalError: DLL problem cause program termination
Failed.
```

Filed this minor issue https://github.com/openstudiocoalition/OpenStudioApplication/issues/234

**Comment by craigers290** [comment]
A few of us tested at VEIC. Did not uncover interface bugs yet but all got the python library error when trying to run models:
Screenshot 2020-09-17 043543

**Comment by jmarrec** [comment]
@craigers290 that's expected, you are using an installer without the fix.

I'm merging this PR to develop to facilitate on-going development, if you uncover problems, please open issues.

**Pull request merged**